lib/Sema/TypeCheckMacros.cpp

```diff
@@ -1082,3 +1077,43 @@ bool swift::expandAttributes(CustomAttr *attr, MacroDecl *macro, Decl *member) {
 
   return addedAttributes;
 }
+
+MacroDecl *
+ResolveAttachedMacroRequest::evaluate(Evaluator &evaluator,
+                                      CustomAttr *attr,
+                                      DeclContext *dc) const {
+  auto &ctx = dc->getASTContext();
+  llvm::TinyPtrVector<ValueDecl *> macros;
+  findMacroForCustomAttr(attr, dc, macros);
+
+  if (macros.empty())
+    return nullptr;
+
+  // Extract macro arguments from the attribute, or create an empty list.
+  ArgumentList *attrArgs;
+  if (attr->hasArgs()) {
+    attrArgs = attr->getArgs();
+  } else {
+    attrArgs = ArgumentList::createImplicit(ctx, {});
+  }
+
+  // Form an `OverloadedDeclRefExpr` with the filtered lookup result above
+  // to ensure @freestanding macros are not considered in overload resolution.
+  FunctionRefKind functionRefKind = FunctionRefKind::SingleApply;
+  auto *identTypeRepr = dyn_cast<IdentTypeRepr>(attr->getTypeRepr());
+  auto macroRefExpr = new (ctx) OverloadedDeclRefExpr(
+      macros, identTypeRepr->getNameLoc(), functionRefKind,
+      /*implicit*/true);
+  auto *call = CallExpr::createImplicit(ctx, macroRefExpr, attrArgs);
+
+  Expr *result = call;
+  TypeChecker::typeCheckExpression(result, dc);
```

Comment:
It is possible for this to fail at this point? If it is should it be handled? Or if it is expect to always succeed do you think assert is worth it?

Reply:
It can fail, in which case diagnostics will be produced and there will not be a `MacroDecl` in the type-checked AST. This is handled in the code below that looks for a `MacroDecl` as the callee, returning `nullptr` and invalidating the attribute if there isn't one.